### ISSUE TITLE
Add support for standard OData error response

### ISSUE THREAD

**Issue by vojtechvit** [object/Simple.OData.Client#96]
Hello again :-)

From my debugging observations, it seems that Simple OData Client doesn't support the [standard JSON OData errors](http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata02/os/odata-json-format-v4.0-errata02-os-complete.html#_Toc403940655) - i.e. the exception thrown on error response (<code>Simple.OData.Client.WebRequestException</code>) does not contain the information contained in the error JSON. It doesn't even contain the response payload - so the error information seems to get lost in the process.

Sample error response (422 Unprocessable Entity):
``` javascript
{
  "error":{
    "code":"4","message":"The inviter does not exist."
  }
}
```

For us the error information is very important. For instance - we have API endpoints that are used as the main backend for user registration forms. Based on error codes/other information returned we decide where in the form (near which field) should we display an error message.

Unless there is a way to retrieve this information which I don't know about, I'd suggest adding some of the following to the library:

1. Make response payload accessible in the exception.
2. Create an extensible error parsing mechanism that would support standard JSON OData errors out of the box and would include this information in the thrown exception. I say extensible because some services may need to add some fields to the error JSON that are non-standard (which is a standard-compliant extension) - and without extensibility the default error class may not be flexible enough.

**Comment by object** [comment]
Great suggestion, thanks. The reason it's handled this way is the library was originally written for OData V3 and there errors were handled differently. I will look into this.

**Comment by vojtechvit** [comment]
:+1:

**Comment by object** [comment]
I have release the version 4.4 with revised implementation of WebRequestException. Now it copies to its Response property the actual HTTP response content. Please check it out.

**Comment by vojtechvit** [comment]
Hi Vagif, I checked and it works ok - with some extension method or helper class it should be easy to parse the error JSON.

However, in the test console application I had to catch an <code>AggregateException</code> - it appears that there is double <code>AggregateException</code> wrapping happening. Could you please fix this so that it's possible to catch the real exception type when using <code>await</code> instead?

Code sample:
``` c#
        private static async Task CreateInvitation()
        {
            try
            {
                var client = new ODataClient("http://localhost:15407/v1");

                var invitation = await client
                    .For("invitations")
                    .Set(new
                    {
                        emailAddress = "my@email.com",
                        inviter = new { id = -1 }
                    })
                    .InsertEntryAsync();

                Console.WriteLine("And here's the token: '{0}'! Magic!", invitation["token"]);
            }
            catch (AggregateException exception)
            {
                var innerException = exception.InnerException as WebRequestException;
                Console.WriteLine("StatusCode: {0}", innerException.Code);
                Console.WriteLine("Content: {0}", innerException.Response);
            }
        }
```

**Comment by object** [comment]
It has to be AggregateException due to the nature of async communication. But of course it shouldn't be wrapped twice. If your example works, then it doesn't look like there are two AggregateException wrapper. Can you give me more details about scenario when you get double wrapper? I don't get it in my tests.

**Comment by vojtechvit** [comment]
The reason why in the aforementioned code sample there is just one Aggregate Exception is because the "await" keyword internally already automatically unwraps one layer of AggregateException. It's a feature of this keyword to allow catching exceptions easier.

So, if you try this code instead, you'll see what I mean:

``` c#
        private static void CreateInvitation()
        {
            try
            {
                var client = new ODataClient("http://localhost:15407/v1");

                var invitation = client
                    .For("invitations")
                    .Set(new
                    {
                        emailAddress = "my@email.com",
                        inviter = new { id = -1 }
                    })
                    .InsertEntryAsync().Result;

                Console.WriteLine("And here's the token: '{0}'! Magic!", invitation["token"]);
            }
            catch (AggregateException exception)
            {
                var innerException = exception.InnerException as AggregateException;
                var innerInnerException = innerException .InnerException as WebRequestException;
                Console.WriteLine("StatusCode: {0}", innerInnerException .Code);
                Console.WriteLine("Content: {0}", innerInnerException .Response);
            }
        }
```

**Comment by object** [comment]
I see, thanks for the clarification, I will look into this.

**Comment by object** [comment]
I think the issue with nested AggregateException is fixed. It will be a part of the version 4.5 that will be out sometime next week, but you can check out its pre-release from MyGet feed:

https://www.myget.org/feed/Packages/simple-odata-client

**Comment by vojtechvit** [comment]
Yes, it works! Great work! ;) Now this code executes correctly:

``` c#
try
{
    var client = new ODataClient("https://localhost:15407/v1");

    var result = await client.For("invitations")
        .Set(new { emailAddress = "some@address@.com" })
        .InsertEntryAsync();
}
catch (WebRequestException exception)
{
    Console.WriteLine(exception.Response);
}
```

**Comment by object** [comment]
Fixed in 4.5.0.
